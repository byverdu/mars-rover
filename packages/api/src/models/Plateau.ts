import * as mongoose from 'mongoose';
import { IPlateau } from './Interfaces';

const PlateauSchema: mongoose.Schema = new mongoose.Schema({
  uuid: { type: String, required: true, unique: true },
  name: { type: String, required: true, unique: true },
  size: {
    width: { type: Number, required: true, unique: true },
    height: { type: Number, required: true, unique: true }
  },
  rovers: { type: Array, required: true, unique: true }
});

// Export the model and return your IPlateau interface
export default mongoose.model<IPlateau>('Plateau', PlateauSchema);
