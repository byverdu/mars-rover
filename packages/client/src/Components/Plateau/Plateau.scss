$purple: #5b13ec;
$white: #ffffff;

.plateau {
  justify-content: center;
  position: relative;
  align-items: center;
  display: flex;
  flex-wrap: wrap;
  color: $white;

  .tile {
    width: 100px;
    height: 100px;
    box-shadow: 1px -1px, -1px 1px;
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    background-image: url(https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSWAgH171cZ6dbemPbqWWCch4VJp_xs1gLbi8sj7qesMaNJMrOD);
    background-color: #cd5c5c;
  }
}
