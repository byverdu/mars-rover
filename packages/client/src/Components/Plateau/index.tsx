import React from 'react';
import './Plateau.scss';
import PlateauSquare from '../PlateauSquare';
import Rover from '../Rover';
import { IRover } from '../../Models/Interfaces';

function roverAxisFormatter(position: string) {
  const coords = position.split(' ');

  return `x${coords[0]}:y${coords[1]}`.trim();
}

function gridBuilder(width: number, height: number, rovers: IRover[]) {
  const wrapper = [];
  const rover: IRover = rovers[0];
  const { rawFormat, position } = rover.lastKnownPosition;
  const nextStepsFormatted = rover.stepsToNextPosition.steps.map((step) =>
    roverAxisFormatter(step)
  );

  function rowBuilder(newHeight) {
    for (let i = 0; i < width; i += 1) {
      const axis = `x${i}:y${height - newHeight}`;
      const roverAxisPos = roverAxisFormatter(rawFormat);
      // console.log(nextStepsFormatted, roverAxisPos)
      const isBusy = axis !== roverAxisPos;
      const willBeBusy = nextStepsFormatted.indexOf(axis) !== -1;

      wrapper.push(
        <PlateauSquare
          willBeBusy={willBeBusy}
          key={axis}
          axis={axis}
          empty={isBusy}
        >
          {!isBusy && (
            <Rover
              axis={roverAxisPos}
              direction={position}
              status={rover.status}
            />
          )}
          {willBeBusy && <div>Will be busy</div>}
        </PlateauSquare>
      );
    }
  }
  let counter = 1;
  while (counter <= width) {
    rowBuilder(counter);
    counter += 1;
  }

  // console.log(wrapper, 'Plateau.scss');

  return wrapper;
}

export interface PlateauProps {
  width: number;
  height: number;
  rovers: IRover[];
}

const Plateau: React.FC<PlateauProps> = ({ width, height, rovers }) => (
  <section className="plateau" style={{ width: `${width * 100}px` }}>
    {gridBuilder(width, height, rovers)}
  </section>
);

export default Plateau;
