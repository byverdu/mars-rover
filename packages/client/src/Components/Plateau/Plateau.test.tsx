import React from 'react';
import Plateau, { PlateauProps } from './index';
import enzyme, { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const props: PlateauProps = {
  width: 5,
  height: 5
};

describe('<Plateau />', () => {
  const wrapper = mount(<Plateau {...props} />);

  it('should create a snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should know it\'s width', () => {
    expect((wrapper.props() as PlateauProps).width).toEqual(5);;
  });

  it('should know it\'s height', () => {
    expect((wrapper.props() as PlateauProps).height).toEqual(5);;
  });

  it('should create as many PlateauSquare in relation with it\' height and width', () => {
    expect(wrapper.find('PlateauSquare').length).toEqual(25);;
  });
});
