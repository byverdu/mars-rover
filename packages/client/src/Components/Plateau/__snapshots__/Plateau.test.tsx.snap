// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Plateau /> should create a snapshot 1`] = `
<Plateau
  height={5}
  width={5}
>
  <section
    className="plateau"
  >
    <PlateauSquare
      axis="x0:y4"
      empty={true}
      key="x0:y4"
    >
      <div
        className="tile empty"
        data-axis="x0:y4"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x1:y4"
      empty={true}
      key="x1:y4"
    >
      <div
        className="tile empty"
        data-axis="x1:y4"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x2:y4"
      empty={true}
      key="x2:y4"
    >
      <div
        className="tile empty"
        data-axis="x2:y4"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x3:y4"
      empty={true}
      key="x3:y4"
    >
      <div
        className="tile empty"
        data-axis="x3:y4"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x4:y4"
      empty={true}
      key="x4:y4"
    >
      <div
        className="tile empty"
        data-axis="x4:y4"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x0:y3"
      empty={true}
      key="x0:y3"
    >
      <div
        className="tile empty"
        data-axis="x0:y3"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x1:y3"
      empty={true}
      key="x1:y3"
    >
      <div
        className="tile empty"
        data-axis="x1:y3"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x2:y3"
      empty={true}
      key="x2:y3"
    >
      <div
        className="tile empty"
        data-axis="x2:y3"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x3:y3"
      empty={true}
      key="x3:y3"
    >
      <div
        className="tile empty"
        data-axis="x3:y3"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x4:y3"
      empty={true}
      key="x4:y3"
    >
      <div
        className="tile empty"
        data-axis="x4:y3"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x0:y2"
      empty={true}
      key="x0:y2"
    >
      <div
        className="tile empty"
        data-axis="x0:y2"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x1:y2"
      empty={true}
      key="x1:y2"
    >
      <div
        className="tile empty"
        data-axis="x1:y2"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x2:y2"
      empty={true}
      key="x2:y2"
    >
      <div
        className="tile empty"
        data-axis="x2:y2"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x3:y2"
      empty={true}
      key="x3:y2"
    >
      <div
        className="tile empty"
        data-axis="x3:y2"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x4:y2"
      empty={true}
      key="x4:y2"
    >
      <div
        className="tile empty"
        data-axis="x4:y2"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x0:y1"
      empty={true}
      key="x0:y1"
    >
      <div
        className="tile empty"
        data-axis="x0:y1"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x1:y1"
      empty={true}
      key="x1:y1"
    >
      <div
        className="tile empty"
        data-axis="x1:y1"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x2:y1"
      empty={true}
      key="x2:y1"
    >
      <div
        className="tile empty"
        data-axis="x2:y1"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x3:y1"
      empty={true}
      key="x3:y1"
    >
      <div
        className="tile empty"
        data-axis="x3:y1"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x4:y1"
      empty={true}
      key="x4:y1"
    >
      <div
        className="tile empty"
        data-axis="x4:y1"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x0:y0"
      empty={true}
      key="x0:y0"
    >
      <div
        className="tile empty"
        data-axis="x0:y0"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x1:y0"
      empty={true}
      key="x1:y0"
    >
      <div
        className="tile empty"
        data-axis="x1:y0"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x2:y0"
      empty={true}
      key="x2:y0"
    >
      <div
        className="tile empty"
        data-axis="x2:y0"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x3:y0"
      empty={true}
      key="x3:y0"
    >
      <div
        className="tile empty"
        data-axis="x3:y0"
      />
    </PlateauSquare>
    <PlateauSquare
      axis="x4:y0"
      empty={true}
      key="x4:y0"
    >
      <div
        className="tile empty"
        data-axis="x4:y0"
      />
    </PlateauSquare>
  </section>
</Plateau>
`;
